File: services/payload_recon.py
"""
Payload Recon Service - ดึง Forms และ URL params จาก URLs
ใช้ข้อมูลจาก path_recon (list ของ URLs) มาวิเคราะห์
รองรับ multithread สำหรับ map_forms หลาย URL พร้อมกัน
และรองรับหน้า SPA/hash routes ผ่าน browser-rendered DOM
"""
import re

import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse, parse_qs
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import List, Dict, Any, Optional


def extract_url_params(url: str) -> Optional[Dict[str, str]]:
    """ดึง Parameter จาก URL เปล่าๆ ที่ไม่มี Form (query string only)."""
    parsed = urlparse(url)
    params = parse_qs(parsed.query)
    if params:
        return {k: v[0] for k, v in params.items()}
    return None


def _infer_field_name(tag: Any, index: int) -> str:
    for attr in ("name", "formcontrolname", "id", "placeholder", "aria-label"):
        value = tag.get(attr)
        if value:
            cleaned = re.sub(r"[^A-Za-z0-9_-]+", "_", str(value).strip()).strip("_")
            if cleaned:
                return cleaned
    return f"field_{index}"


def _extract_forms_from_html(source_url: str, html: str) -> List[Dict[str, Any]]:
    soup = BeautifulSoup(html, "html.parser")
    mapped_data: List[Dict[str, Any]] = []
    request_url = source_url.split("#", 1)[0]

    for form in soup.find_all("form"):
        action = form.get("action") or ""
        full_url = urljoin(request_url, action)
        # strip URL fragment (#) — action="#" หรือ action="" หมายถึง current page
        full_url = full_url.split("#")[0] or request_url
        method = (form.get("method") or "get").upper()
        parsed_url = urlparse(full_url)
        query_params = {k: v[0] for k, v in parse_qs(parsed_url.query).items()}
        body_params: Dict[str, str] = {}
        for idx, ipt in enumerate(form.find_all(["input", "textarea", "select"])):
            name = _infer_field_name(ipt, idx)
            body_params[name] = ipt.get("value") or ""
        mapped_data.append(
            {
                "target_action": full_url.split("?")[0],
                "method": method,
                "query_params": query_params,
                "body_params": body_params,
            }
        )

    if mapped_data:
        return mapped_data

    # SPA บางตัวไม่มี <form> แต่มี input/password + button อยู่ใน rendered DOM
    inputs = soup.find_all(["input", "textarea", "select"])
    password_inputs = soup.find_all("input", {"type": "password"})
    if inputs and password_inputs:
        body_params: Dict[str, str] = {}
        for idx, ipt in enumerate(inputs):
            name = _infer_field_name(ipt, idx)
            body_params[name] = ipt.get("value") or ""
        parsed_url = urlparse(request_url)
        query_params = {k: v[0] for k, v in parse_qs(parsed_url.query).items()}
        mapped_data.append(
            {
                "target_action": source_url.split("?")[0],
                "method": "POST",
                "query_params": query_params,
                "body_params": body_params,
            }
        )
    return mapped_data


def _render_with_playwright(url: str, timeout: int = 10) -> Optional[str]:
    try:
        from playwright.sync_api import sync_playwright
    except Exception:
        return None

    try:
        with sync_playwright() as p:
            browser = p.chromium.launch(headless=True)
            page = browser.new_page()
            page.goto(url, wait_until="networkidle", timeout=timeout * 1000)
            page.wait_for_timeout(1500)
            html = page.content()
            browser.close()
            return html
    except Exception:
        return None


def map_forms(
    url: str,
    timeout: int = 5,
    extra_headers: Optional[Dict[str, str]] = None,
    session: Optional[requests.Session] = None,
) -> List[Dict[str, Any]]:
    """ดึงข้อมูลจาก <form> tags ในหน้า URL นั้น
    extra_headers: ส่ง Cookie/User-Agent เพื่อ fetch หน้า authenticated ได้ถูกต้อง
    session: shared requests.Session เพื่อ keep cookies ไว้ระหว่าง requests (ใช้ cookie jar จริงๆ)
    Returns [] ถ้า 404 หรือถ้าถูก redirect ไป login page (session expire)
    """
    try:
        target_url = url.split("#", 1)[0]
        req = session or requests
        # session มี headers แล้ว (Cookie, User-Agent) — ไม่ต้องส่ง extra_headers ซ้ำเพื่อป้องกัน dup
        hdrs = {} if (session is not None) else (extra_headers or {})
        response = req.get(target_url, timeout=timeout, headers=hdrs, allow_redirects=True)

        # ข้าม 404
        if response.status_code == 404:
            return []

        # ใช้ URL จริงหลัง redirect เป็นฐาน resolve form action
        final_url = response.url or target_url

        # ถ้า redirect ไปหน้า login → session expire → ข้ามหน้านี้
        orig_path = urlparse(target_url).path.lower().rstrip("/")
        final_path = urlparse(final_url).path.lower().rstrip("/")
        if orig_path != final_path and any(x in final_path for x in ("/login", "/signin", "/auth")):
            return []

        html = response.text or ""
        mapped_data = _extract_forms_from_html(final_url, html)
        if mapped_data:
            return mapped_data

        # ถ้าเป็น route แบบ #/login หรือหน้า JS-heavy ให้ลอง render DOM จริง
        if "#/" in url or "ng-version" in html or "<app-root" in html:
            rendered_html = _render_with_playwright(url, timeout=max(timeout, 10))
            if rendered_html:
                mapped_data = _extract_forms_from_html(final_url, rendered_html)
        return mapped_data
    except Exception:
        return []


def _process_one_path(
    path: str,
    extra_headers: Optional[Dict[str, str]] = None,
    session: Optional[requests.Session] = None,
) -> tuple[str, List[Dict[str, Any]], Optional[Dict[str, str]]]:
    """Process one URL: return (path, forms, url_params)."""
    forms = map_forms(path, extra_headers=extra_headers, session=session)
    params = extract_url_params(path)
    return (path, forms, params)


def run_payload_recon(
    paths: List[str],
    max_workers: int = 8,
    extra_headers: Optional[Dict[str, str]] = None,
) -> tuple[Dict[str, Any], Dict[str, Any]]:
    """
    รับ list ของ paths (จาก Katana) แล้วจัดกลุ่มเป็น:
    - grouped_forms: HTML forms (method|action|params)
    - url_entry_points: URL ที่มี query params แต่ไม่มี form
    ใช้ multithread เรียก map_forms แยกต่อ path
    """
    grouped_forms: Dict[str, Any] = {}
    url_entry_points: Dict[str, Any] = {}
    # base_path -> merged params + example_urls
    inferred_by_base: Dict[str, Any] = {}

    # ใช้ shared session เพื่อ keep cookies ระหว่าง requests ต่างๆ — สำคัญสำหรับ authenticated crawl
    # ใช้ headers["Cookie"] โดยตรง (ไม่ใช้ cookie jar) เพราะ jar ต้องการ domain match
    shared_session = requests.Session()
    if extra_headers:
        shared_session.headers.update(extra_headers)

    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        future_to_path = {executor.submit(_process_one_path, p, extra_headers, shared_session): p for p in paths}
        for future in as_completed(future_to_path):
            try:
                path, forms_found, u_params = future.result()
            except Exception:
                continue
            # 1. HTML Forms
            for f in forms_found:
                q_keys = sorted((f.get("query_params") or {}).keys())
                b_keys = sorted((f.get("body_params") or {}).keys())
                sig = f"{f['method']}|{f['target_action']}|Q{q_keys}|B{b_keys}"
                if sig not in grouped_forms:
                    grouped_forms[sig] = {"details": f, "paths": set()}
                grouped_forms[sig]["paths"].add(path)
            # 2. URL params (inferred)
            if u_params:
                base_path = path.split("?")[0]
                if base_path not in inferred_by_base:
                    inferred_by_base[base_path] = {"params": {}, "example_urls": set()}
                # merge params (ไม่ overwrite ถ้ามี key ซ้ำ)
                for k, v in u_params.items():
                    inferred_by_base[base_path]["params"].setdefault(k, v)
                inferred_by_base[base_path]["example_urls"].add(path)

    # 3) Combine: ถ้า base_path เดียวกันมีทั้ง form และ inferred query ให้รวมเข้า form แล้วลบออกจาก query-only list
    used_bases: set[str] = set()
    for sig, data in grouped_forms.items():
        details = data["details"]
        base = details.get("target_action")
        if not base:
            continue
        inferred = inferred_by_base.get(base)
        if not inferred:
            continue
        # merge query params
        q = details.get("query_params") or {}
        for k, v in inferred["params"].items():
            q.setdefault(k, v)
        details["query_params"] = q
        # เก็บ sample URLs ไว้โชว์ได้
        details["example_urls"] = sorted(list(inferred["example_urls"]))[:5]
        used_bases.add(base)

    # 4) Build query-only entry points (เฉพาะที่ไม่มี form)
    for base_path, info in inferred_by_base.items():
        if base_path in used_bases:
            continue
        params = info["params"]
        param_sig = f"GET|{base_path}|{sorted(params.keys())}"
        url_entry_points[param_sig] = {
            "base_path": base_path,
            "params": params,
            "example_urls": info["example_urls"],
        }

    return grouped_forms, url_entry_points

File: backend/services/payload-recon.js
const { URL } = require('url');
const cheerio = require('cheerio');

/**
 * Extract query params from a URL (?a=1&b=2) as flat key->firstValue map.
 */
function extractUrlParams(rawUrl) {
  try {
    const u = new URL(rawUrl);
    const out = {};
    for (const [k, v] of u.searchParams.entries()) {
      if (!(k in out)) out[k] = v;
    }
    return Object.keys(out).length ? out : null;
  } catch {
    return null;
  }
}

function inferFieldName($el) {
  const candidates = [
    'name',
    'formcontrolname',
    'id',
    'placeholder',
    'aria-label',
  ];
  for (const attr of candidates) {
    const value = $el.attr(attr);
    if (!value) continue;
    const cleaned = String(value).trim().replace(/[^A-Za-z0-9_-]+/g, '_').replace(/^_+|_+$/g, '');
    if (cleaned) return cleaned;
  }
  return null;
}

function extractFormsFromHtml(sourceUrl, html) {
  const $ = cheerio.load(html);
  const mapped = [];

  const baseUrl = sourceUrl.split('#')[0];

  $('form').each((_, form) => {
    const $form = $(form);
    const actionAttr = $form.attr('action') || '';
    const formId = ($form.attr('id') || '').trim() || null;
    const formName = ($form.attr('name') || '').trim() || null;

    let fullUrl;
    try {
      fullUrl = new URL(actionAttr || baseUrl, baseUrl).toString();
    } catch {
      fullUrl = baseUrl;
    }

    fullUrl = fullUrl.split('#')[0] || baseUrl;
    const method = ($form.attr('method') || 'GET').toUpperCase();

    const u = new URL(fullUrl, baseUrl);
    const queryParams = {};
    for (const [k, v] of u.searchParams.entries()) {
      if (!(k in queryParams)) queryParams[k] = v;
    }

    const bodyParams = {};
    $form.find('input, textarea, select').each((idx, el) => {
      const $el = $(el);
      const name = inferFieldName($el, idx);
      if (!name) return;
      const value = $el.attr('value') || '';
      bodyParams[name] = value;
    });

    mapped.push({
      target_action: u.origin + u.pathname,
      method,
      query_params: queryParams,
      body_params: bodyParams,
      form_id: formId,
      form_name: formName,
    });
  });

  if (mapped.length) return mapped;

  // Fallback: pages with inputs but no <form> (often SPA-style logins)
  const inputs = $('input, textarea, select');
  const passwordInputs = $('input[type="password"]');
  if (inputs.length && passwordInputs.length) {
    const bodyParams = {};
    inputs.each((idx, el) => {
      const $el = $(el);
      const name = inferFieldName($el, idx);
      if (!name) return;
      const value = $el.attr('value') || '';
      bodyParams[name] = value;
    });

    let u;
    try {
      u = new URL(baseUrl);
    } catch {
      u = null;
    }
    const queryParams = {};
    if (u) {
      for (const [k, v] of u.searchParams.entries()) {
        if (!(k in queryParams)) queryParams[k] = v;
      }
    }

    mapped.push({
      target_action: baseUrl.split('?')[0],
      method: 'POST',
      query_params: queryParams,
      body_params: bodyParams,
    });
  }

  return mapped;
}

async function httpGet(url, timeoutMs, extraHeaders) {
  const controller = new AbortController();
  const id = setTimeout(() => controller.abort(), timeoutMs);

  // Use global fetch if available (Node 18+), otherwise fall back to node-fetch.
  const doFetch =
    typeof fetch === 'function'
      ? fetch
      : (...args) =>
          import('node-fetch').then((m) => m.default(...args));

  try {
    const res = await doFetch(url, {
      method: 'GET',
      headers: extraHeaders || {},
      redirect: 'follow',
      signal: controller.signal,
    });
    const text = await res.text();
    return { status: res.status, url: res.url || url, text };
  } catch (err) {
    return { status: 0, url, text: '', error: err };
  } finally {
    clearTimeout(id);
  }
}

async function mapForms(url, { timeout = 5000, extraHeaders } = {}) {
  const targetUrl = url.split('#')[0];
  const resp = await httpGet(targetUrl, timeout, extraHeaders);

  if (resp.status === 404 || !resp.text) return [];

  let finalUrl = resp.url || targetUrl;
  try {
    const orig = new URL(targetUrl);
    const fin = new URL(finalUrl);
    const origPath = orig.pathname.toLowerCase().replace(/\/+$/, '');
    const finalPath = fin.pathname.toLowerCase().replace(/\/+$/, '');
    if (
      origPath !== finalPath &&
      (finalPath.includes('/login') ||
        finalPath.includes('/signin') ||
        finalPath.includes('/auth'))
    ) {
      return [];
    }
  } catch {
    // ignore URL parse errors
  }

  const html = resp.text || '';
  const mapped = extractFormsFromHtml(finalUrl, html);
  return mapped;
}

async function runPayloadRecon(paths, { timeout = 5000, extraHeaders } = {}) {
  const groupedForms = new Map(); // sig -> { details, paths:Set }
  const inferredByBase = new Map(); // basePath -> { params: {}, exampleUrls:Set }

  const uniquePaths = Array.from(new Set(paths.filter((p) => typeof p === 'string' && p)));

  for (const p of uniquePaths) {
    let formsFound = [];
    let uParams = null;
    try {
      // eslint-disable-next-line no-await-in-loop
      formsFound = await mapForms(p, { timeout, extraHeaders });
      uParams = extractUrlParams(p);
    } catch {
      continue;
    }

    for (const f of formsFound) {
      const qKeys = Object.keys(f.query_params || {}).sort();
      const bKeys = Object.keys(f.body_params || {}).sort();
      const sig = `${f.method}|${f.target_action}|FID${f.form_id || ''}|FNM${f.form_name || ''}|Q${qKeys.join(',')}|B${bKeys.join(',')}`;
      if (!groupedForms.has(sig)) {
        groupedForms.set(sig, {
          details: {
            target_action: f.target_action,
            method: f.method,
            query_params: { ...(f.query_params || {}) },
            body_params: { ...(f.body_params || {}) },
            form_id: f.form_id || null,
            form_name: f.form_name || null,
          },
          paths: new Set(),
        });
      }
      groupedForms.get(sig).paths.add(p);
    }

    if (uParams) {
      const basePath = p.split('?')[0];
      if (!inferredByBase.has(basePath)) {
        inferredByBase.set(basePath, {
          params: {},
          exampleUrls: new Set(),
        });
      }
      const info = inferredByBase.get(basePath);
      for (const [k, v] of Object.entries(uParams)) {
        if (!(k in info.params)) info.params[k] = v;
      }
      info.exampleUrls.add(p);
    }
  }

  const usedBases = new Set();
  // IMPORTANT: do NOT merge inferred URL query params into ALL forms with same action.
  // Example: wp-login.php?action=lostpassword would pollute the normal login form.
  for (const [, data] of groupedForms.entries()) {
    const details = data.details;
    details.example_urls = Array.from(data.paths).slice(0, 5);
    usedBases.add(details.target_action);
  }

  const out = [];

  for (const [, data] of groupedForms.entries()) {
    const details = data.details;
    out.push({
      found_in: Array.from(data.paths),
      action: details.target_action,
      method: details.method,
      query_string: details.query_params || {},
      payload: details.body_params || {},
      example_urls: details.example_urls || [],
      form_id: details.form_id || null,
      form_name: details.form_name || null,
    });
  }

  for (const [basePath, info] of inferredByBase.entries()) {
    if (usedBases.has(basePath)) continue;
    const params = info.params || {};
    out.push({
      found_in: Array.from(info.exampleUrls || []),
      action: basePath,
      method: 'GET',
      query_string: params,
      payload: {},
      example_urls: Array.from(info.exampleUrls || []),
    });
  }

  return out;
}

module.exports = {
  runPayloadRecon,
};

